[region-isolation] Only make partial_apply actor derived if we capture an isolated parameter... not just any actor.

gottesmm opened the pull request:
Before I couldn't do this since, @sil_isolated was not represented on partial_applies. In the first commit in this commit series, I added support to the frontend so that we now emit partial_applies with sil_isolated parameters.

In the second commit, I use this information to restrict when the partial_apply is considered to be "actor derived" to if the partial_apply parameter is @sil_isolated.

rdar://123881277

gottesmm committed:
[concurrency] Make captured self in a closure show up as an @sil_isolated parameter in the closure's function signature.

This ensures that when region isolation is enabled, we can begin properly
modeling when a closure should not be allowed to be isolated. In the next
commit, I am going to loosen this and we should be able to pass in actors
outside of methods without the closure being afflicted with "actor
derived-ness".

gottesmm committed:
[region-isolation] Only make partial_apply actor derived if we capture an isolated parameter... not just any actor.

Before I couldn't do this since, @sil_isolated was not represented on
partial_applies. Since in the previous commit, I added support to the compiler
to represent this, I can now limit this query so now one can pass an actor
instance outside of its method to a nonisolated non-Sendable partial apply.
Since it is Sendable, it is always safe to do this since we are passing the
actor.

rdar://123881277

gottesmm commented:
@swift-ci smoke test

rjmccall commented on lib/SIL/IR/SILFunctionType.cpp:
```diff
@@ -1917,9 +1917,11 @@ lowerCaptureContextParameters(TypeConverter &TC, SILDeclRef function,
 
   for (auto capture : loweredCaptures.getCaptures()) {
     if (capture.isDynamicSelfMetadata()) {
+      assert(!isolatedParam);
```
What's this assertion for?  Is there a reason we can't have a dynamic self metadata capture *and* an isolated parameter or opaque value?  You don't seem to be using it anyway.

gottesmm replied:
I just was being paranoid. I can remove it.

rjmccall reviewed (commented):
Other than the two questionable assertions, this looks good.

gottesmm committed:
[ast] Fix a series of potential nullptr violations in CaptureInfo.

Specifically, CaptureInfo previously as a pattern in CaptureInfo.cpp would
attempt to ascertain if a CapturedValue was a local capture by checking the
CapturedValue's decl. Unfortunately, when we have captured dynamic self
metadata, we do not even have a getDecl() and should return false.

I fixed this by adding a new API to CaptureValue that checks if it has a decl
before checking if the decl is a local capture. This avoids this problem.

gottesmm commented:
Problem was that in CaptureInfo we were determining if a value was isLocalCapture by calling CapturedValue::getDecl(). This doesn't work for dynamic self metadata which is never an isolation parameter (it is a metatype) and doesn't have a decl. I am preparing a small addition to this PR with a patch that fixes that issue by adding a new isLocalCapture to CapturedValue that does the correct thing.

gottesmm committed:
[ast] After determining that the AST has an isolated capture, make sure that we actually process it via an assert.

A little paranoia never hurts.

gottesmm commented:
@swift-ci smoke test

gottesmm committed:
Clean up a little code and add a small note.

gottesmm committed:
Wordsmith a comment.

gottesmm commented:
@swift-ci smoke test

gottesmm committed:
Update some SILGen tests since we now print @sil_isolated.

gottesmm commented:
@swift-ci smoke test

gottesmm committed:
Disable availability checking on a test

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci smoke test macOS platform
